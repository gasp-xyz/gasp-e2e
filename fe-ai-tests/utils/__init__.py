# This file makes the utils directory a Python package
# This allows imports like: from utils.common_test_utils import x
