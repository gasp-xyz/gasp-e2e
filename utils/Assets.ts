import { BN_TEN, BN_THOUSAND } from "@mangata-finance/sdk";
import { ApiPromise } from "@polkadot/api";
import { AccountInfo } from "@polkadot/types/interfaces";
import { BN } from "@polkadot/util";
import { assert } from "console";
import _ from "lodash";
import { MGA_ASSET_ID } from "./Constants";
import { ExtrinsicResult } from "./eventListeners";
import { api, Extrinsic, setupApi, setupUsers } from "./setup";
import { Sudo } from "./sudo";
import { getAssetSupply, getNextAssetId } from "./tx";
import {
  getEventResultFromMangataTx,
  setAssetInfo,
  sudoIssueAsset,
} from "./txHandler";
import { User } from "./User";
import { MangataTypesAssetsCustomMetadata } from "@polkadot/types/lookup";

export class Assets {
  static MG_UNIT: BN = BN_TEN.pow(new BN(18));
  static DEFAULT_AMOUNT = BN_THOUSAND.mul(this.MG_UNIT);

  ///This method create or return the specified number of available assets
  static async getCurrencies(numAssets: number = 2, sudoUser: User) {
    const currencies: string[] = [];
    const numberOfcurrencies = parseInt((await getNextAssetId()).toString());

    if (numAssets > numberOfcurrencies) {
      //we need to create some currencies.
      for (
        let remainingAssetsToCreate = numberOfcurrencies;
        remainingAssetsToCreate < numAssets;
        remainingAssetsToCreate++
      ) {
        await this.issueAssetToSudo(sudoUser);
      }
    }
    //there are some currencies already created.
    for (let index = 0; index < numAssets; index++) {
      await getAssetSupply(new BN(index));
      currencies.push(index.toString());
    }

    return currencies;
  }

  static async setupUserWithCurrencies(
    user: User,
    currencyValues = [new BN(250000), new BN(250001)],
    _sudo: User,
    skipInfo = false
  ): Promise<BN[]> {
    const txs: Extrinsic[] = [];
    await setupApi();
    await setupUsers();
    for (let currency = 0; currency < currencyValues.length; currency++) {
      txs.push(Assets.issueToken(user, currencyValues[currency]));
    }
    const result = await Sudo.batchAsSudoFinalized(...txs);
    const assetIds: BN[] = result
      .filter((X) => X.method === "Issued")
      .map((t) => new BN(t.eventData[0].data.toString()));
    const addInfos: Extrinsic[] = [];
    if (!skipInfo) {
      for (let index = 0; index < assetIds.length; index++) {
        const assetId = assetIds[index];
        addInfos.push(
          Assets.registerAsset(
            `TEST_${assetId}`,
            this.getAssetName(assetId.toString()),
            new BN(18)
          )
        );
      }
      await Sudo.batchAsSudoFinalized(...addInfos);
    }
    user.addAssets(assetIds);
    await user.refreshAmounts();
    return assetIds;
  }

  static async issueAssetToSudo(sudo: User) {
    await this.issueAssetToUser(sudo, new BN(1000), sudo);
  }

  //this method add a certain amount of currencies to a user into a returned currecncyId
  static async issueAssetToUser(
    user: User,
    num = new BN(1000),
    sudo: User,
    skipInfo = false
  ) {
    const result = await sudoIssueAsset(
      sudo.keyRingPair,
      num,
      user.keyRingPair.address
    );
    const eventResult = await getEventResultFromMangataTx(result, [
      "tokens",
      "Issued",
      user.keyRingPair.address,
    ]);

    assert(eventResult.state === ExtrinsicResult.ExtrinsicSuccess);
    const assetId = eventResult.data[0].split(",").join("");
    if (!skipInfo) {
      await setAssetInfo(
        sudo,
        new BN(assetId),
        `TEST_${assetId}`,
        this.getAssetName(assetId),
        `Test token ${assetId}`,
        new BN(18)
      );
    }

    return new BN(assetId);
  }

  static getAssetName(assetID: string) {
    return `m${assetID}`;
  }

  static mintNative(user: User, amount: BN = this.DEFAULT_AMOUNT): Extrinsic {
    user.addAsset(MGA_ASSET_ID);
    return Sudo.sudo(
      api.tx.tokens.mint(MGA_ASSET_ID, user.keyRingPair.address, amount)
    );
  }

  static issueToken(user: User, amount: BN = this.DEFAULT_AMOUNT): Extrinsic {
    return Sudo.sudo(api.tx.tokens.create(user.keyRingPair.address, amount));
  }

  static mintToken(
    asset: BN,
    user: User,
    amount: BN = this.DEFAULT_AMOUNT
  ): Extrinsic {
    return Sudo.sudo(
      api.tx.tokens.mint(asset, user.keyRingPair.address, amount)
    );
  }

  static transfer(target: User, tokenId: BN, amount: BN): Extrinsic {
    return api.tx.tokens.transfer(target.keyRingPair.address, tokenId, amount);
  }

  static transferAll(target: User, tokenId: BN): Extrinsic {
    return api.tx.tokens.transferAll(target.keyRingPair.address, tokenId, true);
  }
  static FinalizeTge(): Extrinsic {
    return Sudo.sudo(api!.tx.issuance.finalizeTge());
  }
  static initIssuance(): Extrinsic {
    return Sudo.sudo(api!.tx.issuance.initIssuanceConfig());
  }
  static promotePool(liquidityId: number, weight: number = 20): Extrinsic {
    return Sudo.sudo(
      api!.tx.proofOfStake.updatePoolPromotion(liquidityId, weight)
    );
  }
  static registerAsset(
    name: string,
    symbol: string,
    decimals: number | BN,
    // location?: MultiLocation,
    location?: object,
    xcmMetadata?: XcmMetadata,
    xykMetadata?: XykMetadata,
    assetId?: number | BN
  ): Extrinsic {
    return Sudo.sudo(
      api.tx.assetRegistry.registerAsset(
        {
          decimals: decimals,
          name: api.createType("Vec<u8>", name),
          symbol: api.createType("Vec<u8>", symbol),
          existentialDeposit: 0,
          location: location ? { V1: location } : null,
          additional: {
            xcm: xcmMetadata,
            xyk: xykMetadata,
          },
        },
        //@ts-ignore
        api.createType("Option<u32>", assetId)
      )
    );
  }

  static updateAsset(assetId: number | BN, update: UpdateAsset): Extrinsic {
    const optional = update.decimals
      ? api.createType("Option<u32>", update.decimals)
      : null;
    return Sudo.sudo(
      api.tx.assetRegistry.updateAsset(
        assetId,
        //@ts-ignore
        optional,
        api.createType("Vec<u8>", update.name),
        api.createType("Vec<u8>", update.symbol),
        null,
        update.location
          ? update.location.location
            ? { V2: update.location } // Some(location)
            : api.createType("Vec<u8>", "0x0100") // Some(None)
          : null, // None
        update.metadata!
      )
    );
  }
  static async disableToken(tokenId: BN) {
    const isRegistered = (
      await api!.query.assetRegistry.metadata(tokenId)
    ).toHuman();
    let extrinsicToDisable: Extrinsic;
    if (!isRegistered) {
      extrinsicToDisable = Assets.registerAsset(
        "testAsset" + tokenId.toString(),
        "testSymbol",
        18,
        undefined,
        undefined,
        {
          operationsDisabled: true,
        },
        tokenId
      );
    } else {
      const metadata: MangataTypesAssetsCustomMetadata = api.createType(
        "MangataTypesAssetsCustomMetadata",
        {
          xyk: {
            operationsDisabled: true,
          },
        }
      );
      extrinsicToDisable = Assets.updateAsset(tokenId, {
        metadata: metadata,
      });
    }
    return await Sudo.asSudoFinalized(extrinsicToDisable);
  }
  static async enableToken(tokenId: BN) {
    const isRegistered = (
      await api!.query.assetRegistry.metadata(tokenId)
    ).toHuman();
    let extrinsicToTokenEnable: Extrinsic;
    if (
      isRegistered &&
      _.get(isRegistered, "additional.xyk.operationsDisabled")
    ) {
      const metadata: MangataTypesAssetsCustomMetadata = api.createType(
        "MangataTypesAssetsCustomMetadata",
        {
          xyk: {
            operationsDisabled: false,
          },
        }
      );
      extrinsicToTokenEnable = await Assets.updateAsset(tokenId, {
        metadata: metadata,
      });
      await Sudo.asSudoFinalized(extrinsicToTokenEnable);
    }
  }
}

export const balance = async (api: ApiPromise, address: string) => {
  const account = await api.query.system.account<AccountInfo>(address);
  return account.data.toJSON();
};

interface XcmMetadata {
  feePerSecond: number;
}

interface XykMetadata {
  operationsDisabled: boolean;
}

// API expects Option<Option<Location>>, we need to wrap it so we can pass
// undefined in UpdateAsset - no update
// { location: undefined } - update with no location
// { location: { ... } } - update with some location
interface UpdateLocation {
  location?: object;
}

// if the value is undefined, it will not be updated
interface UpdateAsset {
  name?: string;
  symbol?: string;
  decimals?: number;
  // location?: MultiLocation,
  location?: UpdateLocation;
  metadata?: MangataTypesAssetsCustomMetadata;
}
