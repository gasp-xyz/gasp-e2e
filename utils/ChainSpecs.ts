import { BN_TEN, BN_ZERO } from "@mangata-finance/sdk";
import { BN, BN_HUNDRED, BN_THOUSAND } from "@polkadot/util";
import { ApiPromise } from "@polkadot/api";

export const TRANSFER_INSTRUCTIONS = 4;
export const WEIGHT_IN_SECONDS = new BN(1_000_000_000_000);

export interface ChainSpec {
  unitCostWeight: number;
  parachain: number;
  assets: Map<AssetSpec, AssetMeta>;
  foreign: Map<AssetSpec, AssetMeta>;
}

export interface AssetSpec {
  symbol: string;
  decimals: number;
  location: object;
  unit: BN;
}

interface AssetMeta {
  fps: BN;
  ed: BN;
  location: object;
}

export enum ChainId {
  Mg,
  Tur,
  Bifrost,
  Statemine,
}

export class AssetId {
  static Tur: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };
  static TurV3: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };

  static Mgx: AssetSpec = {
    symbol: "MGX",
    decimals: 18,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2110 }, { GeneralKey: "0x00000000" }] },
    },
    unit: BN_TEN.pow(new BN(18)),
  };
  static async getLocationFromChain(symbol: string, api: ApiPromise) {
    const entries = await api.query.assetRegistry.metadata.entries();
    const asset = entries.filter((entry) => {
      const elem = entry[1];
      const name = api.createType(
        "Vec<u8>",
        JSON.parse(JSON.stringify(elem)).symbol,
      );
      return name.toHuman() === symbol;
    })[0];
    return JSON.parse(JSON.stringify(asset[1].toHuman())).location;
  }
  static Bnc: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2001 }, { GeneralKey: "0x0001" }] },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static ImbueBncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 0,
      interior: {
        X1: [
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static BncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: 2001 },
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };

  static USDt: AssetSpec = {
    symbol: "USDt",
    decimals: 6,
    location: {
      parents: 1,
      interior: {
        X3: [
          { Parachain: 1000 },
          { PalletInstance: 50 },
          { GeneralIndex: 1984 },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(6)),
  };
}

export const ChainSpecs = new Map<ChainId, ChainSpec>([
  [
    ChainId.Tur,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 2114,
      assets: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(416_000_000_000),
            ed: AssetId.Tur.unit.mul(BN_TEN),
            location: AssetId.Tur.location,
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Mg,
    {
      unitCostWeight: 150_000_000,
      parachain: 2110,
      assets: new Map(),
      foreign: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.Tur.location,
          },
        ],
        [
          AssetId.TurV3,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.TurV3.location,
          },
        ],
        [
          AssetId.Bnc,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.Bnc.location,
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.USDt,
          {
            fps: new BN(13440000),
            ed: BN_ZERO,
            location: AssetId.USDt.location,
          },
        ],
      ]),
    },
  ],
  [
    ChainId.Bifrost,
    {
      unitCostWeight: 200_000_000,
      parachain: 2001,
      assets: new Map([
        [
          AssetId.Bnc,
          {
            fps: new BN(9360000000000),
            ed: AssetId.Bnc.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: { X1: { GeneralKey: "0x0001" } },
            },
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.BncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.ImbueBncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Statemine,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 1000,
      assets: new Map([
        [
          AssetId.USDt,
          {
            fps: new BN(1),
            ed: AssetId.USDt.unit.div(BN_THOUSAND),
            location: {
              parents: 0,
              interior: {
                X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
]);
