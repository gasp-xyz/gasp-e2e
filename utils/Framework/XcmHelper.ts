import {
  Blockchain,
  BuildBlockMode,
  setupWithServer,
} from "@acala-network/chopsticks";
import { StorageValues } from "@acala-network/chopsticks/lib/utils/set-storage";
import { ApiPromise, WsProvider } from "@polkadot/api";
import { HexString } from "@polkadot/util/types";
import { getPort } from "get-port-please";
import * as fs from "fs";
import { bufferToU8a, u8aToHex } from "@polkadot/util";
import { Assets } from "../Assets";
import { alice } from "../setup";
import { Sudo } from "../sudo";

export type SetupOption = {
  endpoint: string;
  blockNumber?: number;
  blockHash?: HexString;
  wasmOverride?: string;
  db?: string;
  types?: Record<string, any>;
  localPort?: number;
  buildBlockMode?: BuildBlockMode;
};

export type DevApi = {
  newBlock: (param?: { count?: number; to?: number }) => Promise<string>;
  setStorage: (values: StorageValues, blockHash?: string) => Promise<void>;
  timeTravel: (date: string | number) => Promise<number>;
  setHead: (hashOrNumber: string | number) => Promise<void>;
};

export type ApiContext = {
  uri: string;
  chain: Blockchain;
  ws: WsProvider;
  api: ApiPromise;
  dev: DevApi;
  teardown: () => Promise<void>;
};

export const setupContext = async ({
  endpoint,
  blockNumber,
  blockHash,
  wasmOverride,
  db,
  types,
  localPort,
  buildBlockMode,
}: SetupOption): Promise<ApiContext> => {
  // random port
  const port = localPort ? localPort : await getPort();
  const config = {
    endpoint,
    port,
    block: blockNumber || blockHash,
    mockSignatureHost: true,
    "build-block-mode": buildBlockMode ? buildBlockMode : BuildBlockMode.Manual,
    db,
    "wasm-override": wasmOverride,
    "registered-types": { types: types },
    "runtime-log-level": 5,
    runtimeLogLevel: 5,
  };
  const { chain, listenPort, close } = await setupWithServer(config);
  const uri = `ws://localhost:${listenPort}`;
  const ws = new WsProvider(uri);
  const api = await ApiPromise.create({
    provider: ws,
    types: types,
  });

  await api.isReady;

  return {
    uri,
    chain,
    ws,
    api,
    dev: {
      newBlock: (param?: { count?: number; to?: number }): Promise<string> => {
        return ws.send("dev_newBlock", [param]);
      },
      setStorage: (values: StorageValues, blockHash?: string) => {
        return ws.send("dev_setStorage", [values, blockHash]);
      },
      timeTravel: (date: string | number) => {
        return ws.send<number>("dev_timeTravel", [date]);
      },
      setHead: (hashOrNumber: string | number) => {
        return ws.send("dev_setHead", [hashOrNumber]);
      },
    },
    async teardown() {
      await api.disconnect();
      await close();
    },
  };
};
export async function upgradeMangata(mangata: ApiContext) {
  const path = `test/xcm/_releasesUT/0.30.0/mangata_kusama_runtime-0.30.0.RC.compact.compressed.wasm`;
  const wasmContent = fs.readFileSync(path);
  const hexHash = mangata.api!.registry.hash(bufferToU8a(wasmContent)).toHex();
  await Sudo.batchAsSudoFinalized(Assets.mintNative(alice));
  await Sudo.asSudoFinalized(
    Sudo.sudo(
      //@ts-ignore
      mangata.api!.tx.parachainSystem.authorizeUpgrade(hexHash)
    )
  );
  const wasmParam = Uint8Array.from(wasmContent);
  const hex = u8aToHex(wasmParam);
  const param = hex.toString();
  await mangata.api.tx.sudo
    .sudo(mangata.api.tx.parachainSystem.enactAuthorizedUpgrade(param))
    .signAndSend(alice.keyRingPair);

  await mangata.dev.newBlock();
  await mangata.dev.newBlock();
}
