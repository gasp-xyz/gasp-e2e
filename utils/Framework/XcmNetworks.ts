import dotenv from "dotenv";
import { setupContext, SetupOption } from "./XcmHelper";

dotenv.config();

const endpoints = {
  kusama: "wss://kusama-rpc.polkadot.io",
  statemine: "wss://ksm-rpc.stakeworld.io/assethub",
  imbue: "wss://kusama.imbuenetwork.com",
  karura: "wss://karura-rpc-0.aca-api.network",
  mangata: "wss://kusama-archive.mangata.online",
  bifrost: "wss://bifrost-parachain.api.onfinality.io/public-ws",
  turing: "wss://rpc.turing.oak.tech",
  moonriver: "wss://moonriver-rpc.dwellir.com",
};

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined;
  }

  return Number(value);
};

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  kusama: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.KUSAMA_WASM,
      blockNumber: toNumber(process.env.KUSAMA_BLOCK_NUMBER),
      endpoint: process.env.KUSAMA_ENDPOINT ?? endpoints.kusama,
      db: process.env.DB_PATH,
      ...options,
    }),
  imbue: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.IMBUE_WASM,
      blockNumber: toNumber(process.env.IMBUE_BLOCK_NUMBER),
      endpoint: process.env.IMBUE_ENDPOINT ?? endpoints.imbue,
      db: process.env.DB_PATH,
      ...options,
    }),
  turing: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.TURING_WASM,
      blockNumber: toNumber(process.env.TURING_BLOCK_NUMBER),
      endpoint: process.env.TURING_ENDPOINT ?? endpoints.turing,
      db: process.env.DB_PATH,
      ...options,
    }),
  statemine: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.STATEMINE_WASM,
      blockNumber: toNumber(process.env.STATEMINE_BLOCK_NUMBER),
      endpoint: process.env.STATEMINE_ENDPOINT ?? endpoints.statemine,
      db: process.env.DB_PATH,
      ...options,
    }),
  karura: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.KARURA_WASM,
      blockNumber: toNumber(process.env.KARURA_BLOCK_NUMBER),
      endpoint: process.env.KARURA_ENDPOINT ?? endpoints.karura,
      db: process.env.DB_PATH,
      ...options,
    }),
  biforst: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.BIFROST_WASM,
      blockNumber: toNumber(process.env.BIFROST_BLOCK_NUMBER),
      endpoint: process.env.BIFROST_ENDPOINT ?? endpoints.bifrost,
      db: process.env.DB_PATH,
      ...options,
    }),
  moonriver: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.MOONRIVER_WASM,
      blockNumber: toNumber(process.env.MOONRIVER_BLOCK_NUMBER),
      endpoint: process.env.MOONRIVER_ENDPOINT ?? endpoints.moonriver,
      db: process.env.DB_PATH,
      ...options,
    }),
  mangata: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.MANGATA_WASM,
      blockNumber: toNumber(process.env.MANGATA_BLOCK_NUMBER),
      endpoint: process.env.MANGATA_ENDPOINT ?? endpoints.mangata,
      db: process.env.DB_PATH,
      types: {
        ShufflingSeed: {
          seed: "H256",
          proof: "H512",
        },
        Header: {
          parentHash: "Hash",
          number: "Compact<BlockNumber>",
          stateRoot: "Hash",
          extrinsicsRoot: "Hash",
          digest: "Digest",
          seed: "ShufflingSeed",
          count: "BlockNumber",
        },
      },
      ...options,
    }),
};
