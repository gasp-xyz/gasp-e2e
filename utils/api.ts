import { ApiPromise } from "@polkadot/api";
import { testLog } from "./Logger";
import { getEnvironmentRequiredVars, getMangataApiUrlPort } from "./utils";
import { Mangata, MangataInstance } from "@mangata-finance/sdk";
import { getPort } from "get-port-please";
import XcmNetworks from "./Framework/XcmNetworks";
import { BuildBlockMode } from "@acala-network/chopsticks";
import { ApiContext } from "./Framework/XcmHelper";

export let api: ApiPromise | null = null;
export let mangata: MangataInstance | null = null;
export let mangataChopstick: ApiContext | null = null;
export let chopstickUri: String;
export const getApi = () => {
  if (!api) {
    throw new Error("Api not initialized");
  }
  return api;
};

export const initApi = async (uri = "") => {
  if (!uri) {
    const { chainUri: envUri } = getEnvironmentRequiredVars();
    uri = envUri;
  }
  if (process.env.CHOPSTICK_ENABLED) {
    const mgaPort = getMangataApiUrlPort();
    const chopstickPort = await getPort();
    mangataChopstick = await XcmNetworks.mangata({
      localPort: chopstickPort,
      buildBlockMode: BuildBlockMode.Instant,
    });
    uri = uri.replace(mgaPort.toString(), chopstickPort.toString());
    chopstickUri = uri;
  }
  testLog.getLog().info(`TEST_INFO: Running test in ${uri}`);
  mangata = Mangata.instance([uri]);
  api = await mangata.api();
  return api;
};

export async function getMangataInstance(uri = ""): Promise<MangataInstance> {
  if (!api) {
    await initApi(uri);
  }
  return mangata!;
}
export const disconnect = async () => {
  //TODO:sdkV2
  //  if (mangata) {
  //    await mangata.disconnect();
  //  } else {
  //    await Mangata.instance([uri]).disconnect();
  //  }
};
