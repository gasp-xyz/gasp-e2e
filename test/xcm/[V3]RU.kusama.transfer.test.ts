import { connectVertical } from "@acala-network/chopsticks";
import { KeyringPair } from "@polkadot/keyring/types";
import { balance } from "../../utils/Assets";
import { AssetId } from "../../utils/ChainSpecs";
import { ApiContext } from "../../utils/Framework/XcmHelper";
import XcmNetworks from "../../utils/Framework/XcmNetworks";
import { devTestingPairs, setupApi, setupUsers } from "../../utils/setup";
import { mangataChopstick } from "../../utils/api";
import {
  expectEvent,
  expectExtrinsicSuccess,
  expectJson,
  matchEvents,
} from "../../utils/validators";
import { BN_BILLION, Mangata } from "@mangata-finance/sdk";
import { BN } from "@polkadot/util";
import { Codec } from "@polkadot/types/types";

/**
 * @group xcm
 * @group proxied
 */
describe("XCM tests for Mangata <-> Kusama", () => {
  let kusama: ApiContext;
  let mangata: ApiContext;
  let alice: KeyringPair;

  beforeAll(async () => {
    await setupApi();
    await setupUsers();
    kusama = await XcmNetworks.kusama();
    mangata = mangataChopstick!;
    await connectVertical(kusama.chain, mangata.chain);
    alice = devTestingPairs().alice;
    await mangata.dev.setStorage({
      Tokens: {
        Accounts: [
          [[alice.address, { token: 4 }], { free: 10 * 1e12 }],
          [
            [alice.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    });
    // await upgradeMangata(mangata);
  });

  beforeEach(async () => {
    await mangata.dev.setStorage({
      Tokens: {
        Accounts: [
          [[alice.address, { token: 4 }], { free: 10 * 1e12 }],
          [
            [alice.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    });
    await kusama.dev.setStorage({
      System: {
        Account: [
          [[alice.address], { providers: 1, data: { free: 10 * 1e12 } }],
        ],
      },
    });
  });

  it("mangata transfer assets to kusama", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdrawKsm({
      account: alice,
      amount: new BN(1e12),
      kusamaAddress: alice.address,
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any[] as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });
    await mangata.chain.newBlock();

    await kusama.chain.newBlock();
    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 4),
    ).toMatchSnapshot();

    expect(await balance(kusama.api, alice.address)).toMatchSnapshot();
    expectEvent(await kusama.api.query.system.events(), {
      event: expect.objectContaining({
        method: "Processed",
        section: "messageQueue",
        data: {
          id: "0xb31883a9f5588fc609fb29ff960cd525ba8dd2846ed40c34418256a591b1d13c",
          origin: {
            Ump: {
              Para: "2,110",
            },
          },
          success: true,
          weightUsed: expect.anything(),
        },
      }),
    });
  });

  it("Kusama transfer assets to mangata", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.depositFromKusama({
      account: alice,
      url: kusama.uri,
      assets: {
        V3: [
          {
            id: { Concrete: { parents: 0, interior: "Here" } },
            fun: { Fungible: 1e12 },
          },
        ],
      },
      beneficiary: {
        V3: {
          parents: 0,
          interior: {
            X1: {
              AccountId32: {
                network: undefined,
                id: alice.addressRaw,
              },
            },
          },
        },
      },
      destination: {
        V3: {
          parents: 0,
          interior: {
            X1: { Parachain: 2110 },
          },
        },
      },
      feeAssetItem: 0,
      weightLimit: "Unlimited",
      txOptions: {
        async extrinsicStatus(events) {
          await matchEvents(events as any as Codec[], "xcmPallet");
        },
      },
    });

    await kusama.chain.newBlock();

    expect(await balance(kusama.api, alice.address)).toMatchSnapshot();
    //TODO: Somehow I can not get the events from the dcmp.
    //    const hashBef = await mangata.api.rpc.chain.getBlockHash(
    //      await getBlockNumber()
    //    );
    await mangata.chain.newBlock();
    // Lets validate balances. Should be enough I guess.
    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 4),
    ).toMatchSnapshot();

    //    const hashAft = await mangata.api.rpc.chain.getBlockHash(
    //      await getBlockNumber()
    //    );
    //    await sleep(3000);
    //    const events1 = await (
    //      await mangata.api.at(hashBef.toString())
    //    ).query.system.events();
    //
    //    const events2 = await (
    //      await mangata.api.at(hashAft.toString())
    //    ).query.system.events();
    //
    //    const events3 = await mangata.api.query.system.events();
    //    await sleep(10000000);
    //
    //    await matchSystemEventsAt(
    //      mangata,
    //      hashBef.toString(),
    //      "parachainSystem",
    //      "dmpQueue"
    //    );
  });
});
