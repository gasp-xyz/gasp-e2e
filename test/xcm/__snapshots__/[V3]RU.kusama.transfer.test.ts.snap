// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCM tests for Mangata <-> Kusama Kusama transfer assets to mangata 1`] = `
{
  "flags": "0x80000000000000000000000000000000",
  "free": 8999347895660,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Kusama Kusama transfer assets to mangata 2`] = `
{
  "free": 10999477160000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Kusama mangata transfer assets to kusama 1`] = `
{
  "free": 9000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Kusama mangata transfer assets to kusama 2`] = `
{
  "flags": "0x80000000000000000000000000000000",
  "free": 10999907996044,
  "frozen": 0,
  "reserved": 0,
}
`;
